# Yet Another Data Science Python Tutorial
This is a basic tutorial to teach a beginner the Python needed to perform Data Science.
It assumes some knowledge of getting around a computer and the Internet,
and a basic background in mathematics ideally with some statistics.
By the end you should have learned how to set up a Python environment,
prepare and run a significant Python program,
and be ready to learn and use basic Data Science libraries such as NumPy, Pandas, SciKit Learn, Seaborn and Matplotlib to do real Data Science.

## Data Science

[Data science](https://www.ibm.com/think/topics/data-science) uses mathematics and statistics to enable organizations to extract insights and guide operational decisions from raw data.
It combines mathematics, computing and domain knowledge to solve real-world problems and uncover hidden patterns.
Companies are generating vast amounts of data that data scientists can clean, statistically analyze, transform, and visualize to discover patterns and predict events in the future.
These can be presented as actionable insights to business decision makers.

## Why Python?

Python is the dominant programming language for data science because its simplicity makes it easy to learn and use and it is currently the most popular programming language in the world.
It was created by [Guido van Rossum](https://gvanrossum.github.io/) in 1991, named for the BBC comedy series [Monty Python’s Flying Circus](https://docs.python.org/3/faq/general.html#why-is-it-called-python).
It's usefulness for data science is because of the large ecosystem of open-source third-party packages for Data Science
and the large community of developers who create libraries and tools to make Python easier to use.


# Data
Humans are the third most intelligent species on the planet (after mice and dolphins, as anyone who has seen or read [Hitchhikers Guide to Galaxy](https://hitchhikers.fandom.com/wiki/Dolphins) knows).
Humans are distinguished by their ability to create *symbols*, or abstract signs that mean some concept or object in the world. The oldest human-created object symbols include 67,800 year old [handprints](https://www.nytimes.com/2026/01/22/world/asia/indonesia-handprint-rock-art-research.html) and 45,500 year old painted [animals](https://www.nytimes.com/2021/01/13/science/cave-painting-indonesia.html) found in caves in Indonesia.
20,000 year old animal symbols have been found that include [markings](https://www.itv.com/news/tyne-tees/2023-01-05/amateur-archaeologist-helps-crack-ice-age-cave-art-code) that tracked their mating behavior as an aid to hunters.
The oldest evidence of counting has been found in 44,000 year old bones with notches that may have marked a [lunar cycle](https://afrolegends.com/2019/05/17/the-lebombo-bone-the-oldest-mathematical-artifact-in-the-world/).

Tally marks were a form of concrete counting where symbols formed a one-to-one correspondence with physical objects or concepts. Eventually shepherds used stones to count their flock and [tokens](https://www.thoughtco.com/clay-tokens-mesopotamian-writing-171673) were used in the Middle East to count agricultural goods and livestock.
Over time, the tokens evolved to represent [abstract quantities](https://sites.utexas.edu/dsb/ain-ghazal/tokens-and-writing-the-cognitive-development/) of what was actually counted.
These tokens became marks inscribed on tablets that led to the development of cuneiform as a method of writing letters, words and text.  

[Data](https://www.ibm.com/think/topics/data) consists of symbols for descriptive facts, figures, observations, and other useful information.
These can be electronically stored as numbers, letters, and text for automatic processing in the computer.
The purpose of data science is to inspect, clean, transform, and interpret raw data to support data-driven decisions.

# Basic Python Data Types
Python has four basic built-in types of data:
- Strings: sequences of text characters
- Integers: negative and positive whole numbers, without fractions
- Floats: continuous numbers with fractions
- Booleans: whether a condition or state is true or not

Note in the following that Python code, lines beginning with `#` are comments for descriptive purposes only and are ignored.

## Strings
Strings are dynamic sequences of text characters inside matching single or double quotes.  
These are examples of creating strings.

In [2037]:
# strings can be defined within single quotes
'text string'

'text string'

In [2038]:
# strings can also be defined within double quotes
"text string"

'text string'

In [2039]:
# double quotes can be included in a single-quoted string
'"double quotes" in a single-quoted text string'

'"double quotes" in a single-quoted text string'

In [2040]:
# single quotes can be included in a double-quoted string
"'single quotes' in a double-quoted string"

"'single quotes' in a double-quoted string"

Double or single quotes can be included in a double or single quoted string by preceding them with backslash character `\`.
The backslash doesn't get into the string, is an *escape* character that allows the following character to "escape" it's normal meaning.

The backslash may appear in the printed result string.
The backslash isn't really there, it's included so string is correct if you would cut and pasted the string into code.
It's a little confusing.

In [2041]:
# double quotes can be included in a double-quoted string by preceding them with back quote character '\'
"'single quotes' and \"double quotes\" in a double-quoted text string"

'\'single quotes\' and "double quotes" in a double-quoted text string'

In [2042]:
# similarly for single quotes
'"double quotes" and \'single quotes\' in a single-quoted text string'

'"double quotes" and \'single quotes\' in a single-quoted text string'

In [2043]:
# backslashes can be included by escaping them with another backslash, only one backslash is actually there
'string including a backslash \\'

'string including a backslash \\'

In [2044]:
# strings can be spread over two lines with the back quote character
"string split over \
two lines"

'string split over two lines'

"Raw" strings begin with `r` followed by a single or double quoted string. Between the quotes, single quotes, double quotes, as well as characters escaped with a backslash `\` are not treated as special.
- `r'...'`
- `r"..."`

The backslash appears as a double backslash in the printed string, the first backslash 'escapes' the second backslash. It is actually a single backslash in the string, like above.

In [2045]:
# single quoted raw string
r'backslash \ is ignored in a single quoted raw string'

'backslash \\ is ignored in a single quoted raw string'

"Raw" strings beginning with `r` followed by a triple quotes (single or double quotes) can include single quotes and double quotes without needing backslashes.
Backslashes can be included without escaping them.
- `r'''...'''`
- `r"""..."""`

In [2046]:
# raw string with single quote and backslash included
r'''raw string with 'single quote' included, backslash \ too'''

"raw string with 'single quote' included, backslash \\ too"

In [2047]:
# raw string with double quote and backslash included
r"""raw string with "double quote" included, backslash \ too"""

'raw string with "double quote" included, backslash \\ too'

String characters can be accessed using a position inside square brackets `[]`, using position 0 as the first character.
Characters can be accessed from the end of the string with negative positions.
The string "STRING" can be visualized as:

<table>
    <tr>
        <td style="text-align: center">0</td>
        <td style="text-align: center">1</td>
        <td style="text-align: center">2</td>
        <td style="text-align: center">3</td>
        <td style="text-align: center">4</td>
        <td style="text-align: center">5</td>
    </tr>
    <tr style="background-color: transparent">
        <td style="border: 1px solid black; width: 0.5in; text-align: center">S</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">T</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">R</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">I</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">N</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">G</td>
        <td></td>
    </tr>
    <tr>
        <td style="text-align: center">-6</td>
        <td style="text-align: center">-5</td>
        <td style="text-align: center">-4</td>
        <td style="text-align: center">-3</td>
        <td style="text-align: center">-2</td>
        <td style="text-align: center">-1</td>
    </tr>
</table>



In [2048]:
# "STRING"[0] is the first character
"STRING"[0]

'S'

In [2049]:
# "STRING"[0] is the fourth character
"STRING"[3]

'I'

In [2050]:
# "STRING"[-1] is the last character
"STRING"[-1]

'G'

In [2051]:
# "STRING"[-3] is the third to last character
"STRING"[-3]

'I'

String *slices* can be accessed by referencing two positions in square brackets separated by a colon `:`.
The characters are selected up to, but not including, the second position.

In [2052]:
# "STRING"[0:2] are the first two characters
"STRING"[0:2]

'ST'

In [2053]:
# "STRING"[1:3] are the second character up to, but not including, the fourth character (i.e., characters 1 and 2)
"STRING"[1:3]

'TR'

Slices can include negative positions.
Again, the characters are selected up to, but not including, the second position.

In [2054]:
# "string"[-3:-1] are the third to last up to, but not including, the last character
"STRING"[-3:-1]

'IN'

In [2055]:
# "string"[-6:-2] are the first four characters
"STRING"[-6:-2]

'STRI'

Either the first or second position may be omitted.
- If the first position is omitted, the slice starts at the beginning of the string, up to but not including the second position.
- If the second position is omitted, the slice ends at with the last character of the string.

In [2056]:
# with the first position omitted, "string"[:2] are the first two characters
"STRING"[:2]

'ST'

In [2057]:
# with the second position omitted, "string"[3:] are the last three characters
"STRING"[3:]

'ING'

In [2058]:
# slices with a negative first position omitting the second position, "string"[-4:] are the last four characters
"STRING"[-4:]

'RING'

In [2059]:
# slices omitting the first position with a negative second position, "string"[:-3] are the first three characters.
"STRING"[:-3]

'STR'

There are a variety of operations that can be performed on strings.

In [2060]:
# two strings separated by a space are combined
"text " "string"

'text string'

In [2061]:
# two strings can also be combined with `+`
"text " + "string"

'text string'

In [2062]:
# strings can be multiplied with `*`
"yes! " * 3

'yes! yes! yes! '

In [2063]:
# strings can have the first letter capitalized
"needs To Be In Capitals".capitalize()

'Needs to be in capitals'

In [2064]:
# strings can be centered in a field
"centered".center(20)

'      centered      '

In [2065]:
# strings can be centered with a fill character
"centered".center(20, '-')

'------centered------'

In [2066]:
# count the instances of a substring
"row, row, row your boat".count("row")

3

In [2067]:
# count the instances of a substring starting at a character
"row, row, row your boat".count("row", 4)

2

In [2068]:
# count the instances of a substring with starting and ending positions
"row, row, row your boat".count("row", 9, 14)

1

In [2069]:
# find index of first string character
"motor".index('o')

1

In [2070]:
# find index of first string character starting at a position
"motor".index('o', 2)

3

In [2071]:
# find the position of a substring
"row, row, row your boat".find("row")

0

In [2072]:
# find the position of a substring starting at a character
"row, row, row your boat".find("row", 4)

5

In [2073]:
# find the position of a substring with starting and ending positions
"row, row, row your boat".find("row", 9, 14)

10

In [2074]:
# join string characters separated by a fill character
"-".join('separated')

's-e-p-a-r-a-t-e-d'

In [2075]:
# strings can be left justified in a field
"on the left".ljust(20)

'on the left         '

In [2076]:
# strings can be left justified with a fill character
"on the left".ljust(20, '-')

'on the left---------'

In [2077]:
# strings can be converted to lowercase
"Needs To Be In Lowercase".lower()

'needs to be in lowercase'

In [2078]:
# strings can have leading spaces removed
"   indented".lstrip()

'indented'

In [2079]:
# strings can have any of a set of leading characters removed
"http://www.google.com".lstrip('htp:/')

'www.google.com'

In [2080]:
# strings can have a fixed leading string removed
"four score and seven years ago".removeprefix('four ')

'score and seven years ago'

In [2081]:
# strings can have a fixed ending string removed
"four score and seven years ago".removesuffix(' ago')

'four score and seven years'

In [2082]:
# string substrings can be replaced
"row, row, row your boat".replace('row', 'paddle')

'paddle, paddle, paddle your boat'

In [2083]:
# a fixed number of string substrings can be replaced
"row, row, row your boat".replace('row', 'paddle', 2)

'paddle, paddle, row your boat'

In [2084]:
# find the last position of a substring
"row, row, row your boat".rfind("row")

10

In [2085]:
# find the last position of a substring starting at a character
"row, row, row your boat".find("row", 0)

0

In [2086]:
# find the last position of a substring with starting and ending positions
"row, row, row your boat".find("row", 3, 10)

5

In [2087]:
# strings can be right justified in a field
"on the right".rjust(20)

'        on the right'

In [2088]:
# strings can be right justified with a fill character
"on the right".rjust(20, '-')

'--------on the right'

In [2089]:
# strings can have trailing spaces removed
"right padded   ".rstrip()

'right padded'

In [2090]:
# strings can have any of a set of leading characters removed
"python3.exe".rstrip('.ex')

'python3'

In [2091]:
# strings can have leading and trailing spaces removed
"   in the middle   ".strip()

'in the middle'

In [2092]:
# strings can have any of a set of leading and trailing characters removed
"www.google.com".strip('.wcom')

'google'

In [2093]:
# strings can be converted to title case (each word capitalized)
"hello world".title()

'Hello World'

In [2094]:
# strings can be converted to uppercase
"Needs To Be In Uppercase".upper()

'NEEDS TO BE IN UPPERCASE'

## Integers
Integers are negative and positive whole numbers, without fractions.
The result of any integer operation is always an integer.
As anyone familiar with Hitchhikers Guide to Galaxy would know,
the Answer to the Ultimate Question of Life, the Universe, and Everything is [42](https://hitchhikers.fandom.com/wiki/42).

In [2095]:
# use '+' for integer addition
17 + 25

42

In [2096]:
# use '-' for integer subtraction
66 - 24

42

In [2097]:
# use '*' for integer multiplication
2 * 21

42

In [2098]:
# '//' performs integer division (integer result, remove any fraction)
128 // 3

42

In [2099]:
# use '%' for integer modulo (remainder after division)
128 % 43

42

In [2100]:
# integer absolute value
abs(-42)

42

In [2101]:
# convert to integer
int("42")

42

In [2102]:
# integer exponentiation
42 ** 2

1764

In [2103]:
# convert an integer to a string
str(42)

'42'

In [2104]:
# convert a string to an integer
int("42")

42

## Floats
Floats are negative and positive real numbers with fractions.
The result of any floating operation is always an float.

In [2105]:
# use '+' for floating addition
17.0 + 25

42.0

In [2106]:
# use '-' for floating subtraction
66.0 - 24

42.0

In [2107]:
# use '*' for floating multiplication
2 * 21.0

42.0

In [2108]:
# use '/' for floating division (returns a float even for integers)
126 / 3

42.0

In [2109]:
# use '%' for  modulo (remainder after division)
128.0 % 43

42.0

In [2110]:
# float absolute value
abs(-42.0)

42.0

In [2111]:
# round a float down
round(42.2)

42

In [2112]:
# round a float up
round(41.7)

42

In [2113]:
# convert to float
float("42.0")

42.0

In [2114]:
# exponentiation (here, square root)
1764 ** 0.5

42.0

In [2115]:
# convert a float to a string
str(42.0)

'42.0'

## Booleans

Booleans are truth values, and always have one of the two values True and False.
There are a number of comparison operations that return booleans.
They work for integer, float, as well as string types.
For integers and floats, they work as expected.
For strings, characters are compared from the left to find the result.

| Comparison | Meaning |
| - | - |
| < |  less than |
| <= | less than or equal |
| > |  greater than |
| >= | greater than or equal |
| == | equal |
| != | not equal |


In [2116]:
# string less than
"abc" < "abc"

False

In [2117]:
# string less than or equal
"abc" <= "abc"

True

In [2118]:
# string greater than
"abc" > "abc"

False

In [2119]:
# string greater than or equal
"abc" >= "abc"

True

In [2120]:
# string equal
"abc" == "abc"

True

In [2121]:
# string not equal
"abc" != "abc"

False

Conditions have operators to combine or modify their values.

| operator | example | result |
| - | - | - |
| `and` | *condition* and *condition* | True if both conditions are true  |
| `or` | *condition* or *condition* | True if one of the conditions is true |
| `not` | not *condition* | Reverses the result, False if the condition is true |

The results of the operators can be summarized in a *truth table*.

| a | b | a `and` b | a `or` b | `not` a |
| - | - | - | - | - |
| True | True | True | True | False |
| True | False | False | True | False |
| False | True | False | True | True |
| False | False | False | False | True |

In [2122]:
# a string begins with 'a' (true) and is length 3 (true)
("abc"[0] == 'a') and (len("abc") == 3)

True

In [2123]:
# a string ends with bc' (true) or is longer than 5 (false)
("abc"[-2:] == 'bc') or (len("abc") > 5)

True

In [2124]:
# a string is not 2 long (false, could also have used !=)
not (len("abc") == 2)

True

There are a number of string functions that return booleans.

In [2125]:
# test a string ends with a substring
'Python'.endswith('thon')

True

In [2126]:
# test a string ends with a substring, starting at a position
'Python'.endswith('thon', 3)

False

In [2127]:
# test if all characters in the string are alphanumeric
'abc123'.isalnum()

True

In [2128]:
# test if all characters in the string are decimal characters (0 to 9)
'0123456789'.isdecimal()

True

In [2129]:
# test if all characters in the string are alphabetic
'abcdef'.isalpha()

True

In [2130]:
# test if all characters in the string are spaces
'   '.isspace()

True

In [2131]:
# test if all characters in the string are uppercase
'BANANA'.isupper()

True

In [2132]:
# test a string starts with a substring
'Python'.startswith('Pyth')

True

In [2133]:
# test a string starts with a substring, starting at a  position
'Python'.startswith('Pyth', 2)

False

In [2134]:
# test if a substring is in a string
'ytho' in 'Python'

True

In [2135]:
# test if a substring is not in a string
'ython' not in 'Python'

False

## Operator Precedence

The table below shows the precedence and associativity of arithmetic operators

| Operator | Description | Associativity |
| - | - | - |
| `**` | exponentiation | right to left |
| `%`, `*`, `/`, `//` | modulo, multiplication, division and integer division | left to right |
| `+`, `-` | addition and subtraction | left to right |
| `<`, `<=`, `>`, `>=`, `==`, `!=` | comparisons | left to right |
| `not` | logical operator | left to right |
| `and` | logical operator | left to right |
| `or` | logical operator | left to right |

# Python Data Structures
Python data structs collect group data value. These include:
- Lists: an ordered sequence of any values
- Sets: a group of values with no duplicates
- Dictionaries: a group where values are associated with a key

## Lists
Lists are dynamic sequences of values that can be accessed by position. Lists are formed with enclosing values of any type between square brackets []. List values are accessed by position counting from 0:
- list[0] is the first element
- list[1:3] are the first *up to* the third element (i.e., elements 1 and 2)
- list[-1] is the last element
  
and so on. The list `["text", 1, True, 4.2]` can be visualized as:
<table>
    <tr>
        <td style="text-align: center">0</td>
        <td style="text-align: center">1</td>
        <td style="text-align: center">2</td>
        <td style="text-align: center">3</td>
    </tr>
    <tr style="background-color: transparent">
        <td style="border: 1px solid black; width: 0.5in; text-align: center">"text"</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">1</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">True</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">4.2</td>
        <td></td>
    </tr>
    <tr>
        <td style="text-align: center">-4</td>
        <td style="text-align: center">-3</td>
        <td style="text-align: center">-2</td>
        <td style="text-align: center">-1</td>
    </tr>
</table>


There are a variety of operations that can be performed on lists.  

In [2136]:
# a sample list with mixed types
['text', 1, True, 4.2]

['text', 1, True, 4.2]

In [2137]:
# length of a list
len(['a', 'b', 'c', 'b'])

4

In [2138]:
# select a list value (counting from 0)
['a', 'b', 'c', 'd'][1]

'b'

In [2139]:
# select a list slice of from 1 up to (not including) 3
['a', 'b', 'c', 'd'][1:3]

['b', 'c']

In [2140]:
# select a list slice of from 1 to the end
['a', 'b', 'c', 'd'][1:]

['b', 'c', 'd']

In [2141]:
# select a list slice up to (not including) 2
['a', 'b', 'c', 'd'][:2]

['a', 'b']

In [2142]:
# select every other list value
['a', 'b', 'c', 'd'][0:3:2]

['a', 'c']

In [2143]:
# select last list value
['a', 'b', 'c', 'd'][-1]

'd'

In [2144]:
# find index of first list value
['a', 'b', 'c', 'd', 'b'].index('b')

1

In [2145]:
# count instances of a list value
['a', 'b', 'c', 'd', 'b'].count('b')

2

In [2146]:
# test if a value is in list
'b' in ['a', 'b', 'c', 'd']

True

In [2147]:
# test if a value is not in list
'e' not in ['a', 'b', 'c', 'd']

True

In [2148]:
# repeat a list
['a', 'b', 'c', 'd']*2

['a', 'b', 'c', 'd', 'a', 'b', 'c', 'd']

In [2149]:
# combine two lists
['a', 'b', 'c', 'd'] + ['e', 'f', 'g', 'h']

['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']

In [2150]:
# list minimum value
min([210.1, 81, 42, 115])

42

In [2151]:
# list maximum value
max([20.1, 8, 42, 15])

42

In [2152]:
# sum of list values
sum([12, 9, 13, 8])

42

In [2153]:
# sorted list
sorted([20.1, 1, 42, 15])

[1, 15, 20.1, 42]

In [2154]:
# split a string at spaces into a list
'1 2   3'.split()

['1', '2', '3']

In [2155]:
# split a string at a character into a list
'1,2,3'.split(',')

['1', '2', '3']

In [2156]:
# join list strings separated by a fill character
"-".join(['one', 'two', 'three'])

'one-two-three'

## Sets
Sets are an unordered group of values with no duplicates. Sets have rich set of operations as defined by [set theory](https://realpython.com/python-sets/), of which a useful set are implemented in Python.

In [2157]:
# set without duplicates
{'apple', 'orange', 'apple', 'pear', 'orange', 'banana'}

{'apple', 'banana', 'orange', 'pear'}

In [2158]:
# count the number of values in the set
len({'apple', 'banana', 'orange', 'pear'})

4

In [2159]:
# detecting if a value is in a set
'orange' in {'apple', 'banana', 'orange', 'pear'}

True

In [2160]:
# detecting if a value is not in a set
'grape' not in {'apple', 'banana', 'orange', 'pear'}

True

In [2161]:
# set intersection: fruit in both sets
{'apple', 'banana', 'orange', 'pear'} & {'grape', 'banana', 'peach', 'orange'}

{'banana', 'orange'}

In [2162]:
# set union: fruit in either set
{'apple', 'banana', 'orange', 'pear'} | {'grape', 'banana', 'peach', 'orange'}

{'apple', 'banana', 'grape', 'orange', 'peach', 'pear'}

In [2163]:
# set difference: fruit in one set but not in another
{'apple', 'banana', 'orange', 'pear'} - {'grape', 'banana', 'peach', 'orange'}

{'apple', 'pear'}

In [2164]:
# fruit in one or the othere set but not both
{'apple', 'banana', 'orange', 'pear'} ^ {'grape', 'banana', 'peach', 'orange'}

{'apple', 'grape', 'peach', 'pear'}

In [2165]:
# is one set a subset of another
{'banana', 'pear'} <= {'apple', 'banana', 'orange', 'pear'}

True

In [2166]:
# is one set a superset of another
{'apple', 'banana', 'orange', 'pear'} >= {'banana', 'pear'}

True

## Dictionaries
Dictionaries are a group where you look up a value with another value, a key, like looking up a word definition in a real dictionary. Dictionaries are formed with enclosing pairs of values of any type between braces {}. Dictionary values are accessed by key using square brackets.
The dictionary `{'leaf': 'green', 'sky': 'blue', 'apple': 'red'}` can be visualized as:
<table>
    <tr>
        <th style="text-align: center">key</th>
        <th style="text-align: center">value</th>
    </tr>
    <tr style="background-color: transparent">
        <td style="border: 1px solid black; width: 0.5in; text-align: center">leaf</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">green</td>
    </tr>
    <tr style="background-color: transparent">
        <td style="border: 1px solid black; width: 0.5in; text-align: center">sky</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">blue</td>
    </tr>
    <tr style="background-color: transparent">
        <td style="border: 1px solid black; width: 0.5in; text-align: center">apple</td>
        <td style="border: 1px solid black; width: 0.5in; text-align: center">red</td>
    </tr>
</table>



In [2167]:
# a sample dictionary
{'leaf': 'green', 'sky': 'blue', 'apple': 'red'}

{'leaf': 'green', 'sky': 'blue', 'apple': 'red'}

In [2168]:
# select a dictionary value
{'leaf': 'green', 'sky': 'blue', 'apple': 'red'}['leaf']

'green'

In [2169]:
# count the dictionary pairs
len({'leaf': 'green', 'sky': 'blue', 'apple': 'red'})

3

In [2170]:
#  detect if a key is in a dictionary
'leaf' in {'leaf': 'green', 'sky': 'blue', 'apple': 'red'}

True

In [2171]:
#  detect if a key is not in a dictionary
'cloud' in {'leaf': 'green', 'sky': 'blue', 'apple': 'red'}

False

In [2172]:
# list all the keys in a dictionary. The result of keys() is a special data structure, it needs to be converted to a list.
list({'leaf': 'green', 'sky': 'blue', 'apple': 'red'}.keys())

['leaf', 'sky', 'apple']

In [2173]:
# list all the values in a dictionary. The result of values() is a special data structure, it needs to be converted to a list.
list({'leaf': 'green', 'sky': 'blue', 'apple': 'red'}.values())

['green', 'blue', 'red']

# Python Libraries

The best feature of Python are that it has open-source libraries for almost anything you want a program to do.
Some useful ones are:
- Mathematics (`math`)
- Statistics (`statistics`)
- Regular Expressions (`regex`)
- Data analysis (`pandas`)
- Numerical computing (`numpy`)
- Statistical and mathematical algorithms (`scipy`)
- Data visualization (`matplotlib`)
- Statistical graphics (`seaborn`)
- Machine learning (`scikit-learn`)

To use libraries, we use the `import` statement at the top of the program.
It makes the functions in the library available to Python.

`import` *library*

## [Mathematics](https://docs.python.org/3/library/math.html) - `math` library
Python has a wide variety of standard mathematics functions.
To use it, we import the `math` module.

```import math```

In [2174]:
# remove a fraction from a float
import math
math.trunc(42.2)

42

In [2175]:
# square root
import math
math.sqrt(1764)

42.0

In [2176]:
# pi
import math
math.pi

3.141592653589793

In [2177]:
# e
import math
math.e

2.718281828459045

In [2178]:
# logarithm, base e
import math
math.log(42)

3.7376696182833684

In [2179]:
# logarithm, base 10
import math
math.log(42, 10)

1.6232492903979003

In [2180]:
# convert angle to radians (= degrees * pi / 180)
math.radians(42)

0.7330382858376184

In [2181]:
# cosine of angle in radians
import math
math.cos(math.radians(42))

/usr/local/lib/python3.12/dist-packages/IPython/core/displayhook.py:275: UserWarning: Output cache limit (currently 1000 entries) hit.
Flushing oldest 200 entries.
  warn('Output cache limit (currently {sz} entries) hit.\n'


0.7431448254773942

In [2182]:
# sine of angle in radians
import math
math.sin(math.radians(42))

0.6691306063588582

## [Statistics](https://docs.python.org/3/library/statistics.html#statistics) - `statistics` library
Python has library with mosts standard mathematics functions.
To use it, we import the `statistics` module.

```import statistics```

In [2183]:
# mean
import statistics
data = [15, 21, 28, 33, 40, 44, 49, 56, 62, 72]
statistics.mean(data)

42

In [2184]:
# median
import statistics
data = [42.0, 250.5, 12.4, 42.0, 98.7, 12.4, 42.0, 86.4, 134.8, 3.7, 111.6]
statistics.median(data)

42.0

In [2185]:
# mode
import statistics
data = [42.0, 250.5, 12.4, 42.0, 98.7, 12.4, 42.0, 86.4, 134.8, 3.7, 111.6]
statistics.mode(data)

42.0

In [2186]:
# standard deviation
import statistics
data = [42.0, 250.5, 12.4, 42.0, 98.7, 12.4, 42.0, 86.4, 134.8, 3.7, 111.6]
statistics.stdev(data)

72.51461044005357

## [Date and Time functions](https://docs.python.org/3/library/datetime.html) - `datetime` library
The `datetime` module allows handling dates, times, and durations.
These are handled with a number of datetime `objects`:
- `date`: calendar dates
- `time`: date with time of day
- `timedelta`: durations

In [2187]:
# todays date
import datetime
datetime.date.today()

datetime.date(2026, 8, 20)

In [2188]:
# an arbitrary date
import datetime
datetime.date(2020, 12, 31)

datetime.date(2020, 12, 31)

In [2189]:
# replace the day in a date
import datetime
datetime.date(2020, 12, 31).replace(day=1)

datetime.date(2020, 12, 1)

In [2190]:
# replace the month in a date
import datetime
datetime.date(2020, 12, 31).replace(month=8)

datetime.date(2020, 8, 31)

In [2191]:
# replace the year in a date
import datetime
datetime.date(2020, 12, 31).replace(year=2015)

datetime.date(2015, 12, 31)

In [2192]:
# get the weekeday of today's date (as a number)
import datetime
datetime.date.today().weekday()

3

In [2193]:
# get the current time in 24-hour format (with milliseconds)
import datetime
datetime.datetime.now().time()

datetime.time(21, 3, 27, 495723)

In [2194]:
# get the current hour
import datetime
datetime.datetime.now().time().hour

21

In [2195]:
# get the current minute
import datetime
datetime.datetime.now().time().minute

3

In [2196]:
# get a duration (as days and seconds)
import datetime
datetime.timedelta(
    days=50,
    seconds=27,
    minutes=5,
    hours=8,
    weeks=2
)

datetime.timedelta(days=64, seconds=29127)

In [2197]:
# get a duration as the difference between two times
import datetime
datetime.datetime(2026, 8, 16, 10, 0, 0) - datetime.datetime(2026, 8, 14, 12, 30, 45)

datetime.timedelta(days=1, seconds=77355)

The `datetime` library has functions with a fairly extensive range of options for converting dates to text with `strftime`
and converting text to dates with `strptime`.

| Directive | Meaning | Example |
| - | - | - |
| %a | Weekday abbreviated name. | Sun, Mon, ..., Sat |
| %A | Weekday full name | Sunday, Monday, ..., Saturday |
| %w | Weekday as a decimal number, where 0 is Sunday and 6 is Saturday | 0, 1, ..., 6 |
| %d | Day of the month as a zero-padded decimal number | 01, 02, ..., 31 |
| %b | Month abbreviated name | Jan, Feb, ..., Dec |
| %B | Month full name | January, February, ..., December |
| %m | Month as a zero-padded decimal number | 01, 02, ..., 12 |
| %y | Year without century as a zero-padded decimal number | 00, 01, ..., 99 |
| %Y | Year with century as a decimal number | 0001, 0002, ..., 2013, 2014, ..., 9998, 9999 |
| %H | Hour (24-hour clock) as a zero-padded decimal number | 00, 01, ..., 23 |
| %I | Hour (12-hour clock) as a zero-padded decimal number | 01, 02, ..., 12 |
| %p | AM or PM | AM, PM |
| %M | Minute as a zero-padded decimal number | 00, 01, ..., 59 |
| %S | Second as a zero-padded decimal number | 00, 01, ..., 59 |
| %f | Microsecond as a decimal number, zero-padded to 6 digits | 000000, 000001, ..., 999999 |
| %j | Day of the year as a zero-padded decimal number | 001, 002, ..., 366 |
| %c | Full date and time  | Tue Aug 16 21:30:00 1988 |
| %x | Full date | 08/16/1988 |
| %X | Full time (24 hour) | 21:30:00 |
| %% | A literal '%'character | % |

In [2198]:
# format today as a full date/time
import datetime
datetime.datetime.now().strftime("%c")

'Thu Aug 20 21:03:27 2026'

In [2199]:
# format today date/time like "05/30/2022 04:12:50 AM"
import datetime
datetime.datetime.now().strftime("%m/%d/%Y %I:%M:%S %p")

'08/20/2026 09:03:27 PM'

In [2200]:
# format today date like "09/12/2022"
import datetime
datetime.datetime.now().strftime("%m/%d/%Y")

'08/20/2026'

In [2201]:
# format today date/time like "October 5, 2022"
import datetime
datetime.datetime.now().strftime("%B %d, %Y")

'August 20, 2026'

In [2202]:
# format today date/time like "04-Feb-2022" with 12-hour time
import datetime
datetime.datetime.now().strftime("%d-%b-%Y")

'20-Aug-2026'

In [2203]:
# Translate today date/time like "05/30/2022 04:12:50 AM" to datetime object
import datetime
datetime.datetime.strptime("05/30/2022 04:12:50 AM","%m/%d/%Y %I:%M:%S %p")

datetime.datetime(2022, 5, 30, 4, 12, 50)

In [2204]:
# Translate today date/time like October 5, 2022" to datetime object
import datetime
datetime.datetime.strptime("October 5, 2022", "%B %d, %Y")

datetime.datetime(2022, 10, 5, 0, 0)

## [Regular Expressions](https://docs.python.org/3/library/re.html) - `re` library
The `re` module offers a set of functions that allows us to search a string for a match:

| Function | Description |
| - | - |
| findall | Returns a list containing all matches |
| search | Returns a Match object if there is a match anywhere in the string |
| split | Returns a list where the string has been split at each match |
| sub | Replaces one or many matches with a string |



## Metacharacters
Metacharacters are characters with a special meaning:

<table>
    <tr>
        <th style="border-bottom: 1px solid black;">Character</th>
        <th style="border-bottom: 1px solid black;">Description</th>
        <th style="border-bottom: 1px solid black;">Example</th>
    </tr>    
    <tr style="background-color: transparent">
        <td style="border-bottom: 1px solid black;">[]</td>
        <td  style="border-bottom: 1px solid black;">A set of characters</td>
        <td  style="border-bottom: 1px solid black;">"[a-m]"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">\</td>
        <td  style="border-bottom: 1px solid black;">
            Signals a special sequence (can also be used to escape special characters)
        </td>
        <td  style="border-bottom: 1px solid black;">"\d"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">.</td>
        <td  style="border-bottom: 1px solid black;">Any character</td>
        <td  style="border-bottom: 1px solid black;">"he..o"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">^</td>
        <td  style="border-bottom: 1px solid black;">Starts with</td>
        <td  style="border-bottom: 1px solid black;">"^hello"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">\$</td>
        <td  style="border-bottom: 1px solid black;">Ends with</td>
        <td  style="border-bottom: 1px solid black;">"planet$"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">*</td>
        <td  style="border-bottom: 1px solid black;">Zero or more occurrences</td>
        <td  style="border-bottom: 1px solid black;">"he.*o"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">+</td>
        <td  style="border-bottom: 1px solid black;">One or more occurrences</td>
        <td  style="border-bottom: 1px solid black;">"he.+o"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">?</td>
        <td  style="border-bottom: 1px solid black;">Zero or one occurrences</td>
        <td  style="border-bottom: 1px solid black;">"he.?o"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">{<em>m</em>}</td>
        <td  style="border-bottom: 1px solid black;">Exactly <em>m</em> occurrences</td>
        <td  style="border-bottom: 1px solid black;">"he.{2}o"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">{<em>m</em>,<em>n</em>}</td>
        <td  style="border-bottom: 1px solid black;">Between <em>m</em> and <em>n</em> occurrences</td>
        <td  style="border-bottom: 1px solid black;">"he.{2,3}o"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">|</td>
        <td  style="border-bottom: 1px solid black;">Either or</td>
        <td  style="border-bottom: 1px solid black;">"falls|stays"</td>
    </tr>
    <tr style="background-color: transparent">
        <td  style="border-bottom: 1px solid black;">()</td>
        <td  style="border-bottom: 1px solid black;">Capture and group</td>
        <td  style="border-bottom: 1px solid black;"></td>    
    </tr>
</table>



## Special Sequences
A special sequence is a backslash `\` followed by one of the characters in the list below, where the backslash *escapes* the following character and has a special meaning.
In match patterns, in a single or double quoted string the backslash must be escaped with another backslash.
Simpler is to use a raw string `r'...'` or 'r"..."` that allow a bare backslash as described in the Strings section.

| Character | Description |
| - | - |
| \A | Returns a match if the specified characters are at the beginning of the string |
| \b | Returns a match where the specified characters are at the beginning or at the end of a word |
| \B | Returns a match where the specified characters are present, but NOT at the beginning (or at the end) of a word |
| \d | Returns a match where the string contains digits (numbers from 0-9) |
| \D | Returns a match where the string DOES NOT contain digits |
| \s | Returns a match where the string contains a white space character |
| \S | Returns a match where the string DOES NOT contain a white space character |
| \w | Returns a match where the string contains any word characters (characters from a to Z, digits from 0-9, and the underscore _ character) |
| \W | Returns a match where the string DOES NOT contain any word characters |
| \Z | Returns a match if the specified characters are at the end of the string |


## Sets
A set is a set of characters inside a pair of square brackets [] with a special meaning. Escape a square bracket to include an actual square bracket in a pattern, `\[` or `\]`.

| Set | Description |
| - | - |
| [arn] | Returns a match where one of the specified characters (**a**, **r**, or **n**) is present |
| [a-n] | Returns a match for any lower case character, alphabetically between **a** and **n** |
| [^arn] | Returns a match for any character EXCEPT **a**, **r**, and **n** |
| [0123] | Returns a match where any of the specified digits (**0**, **1**, **2**, or **3**) are present |
| [0-9] | Returns a match for any digit between **0** and **9** |
| [0-5][0-9] | Returns a match for any two-digit numbers from **00** and **59** |
| [a-zA-Z] | Returns a match for any character alphabetically between **a** and **z**, lower case OR upper case |
| [+] | In sets, **+**, *****, **.**, **\|**, **()**, **$**, **{}** has no special meaning, so **[+]** means: return a match for any **+** character in the string |

### The findall() Function
The `findall() function returns a list containing all matches.

In [2205]:
# Print a list of all matches in the order they are found
import re
txt = "The rain in Spain"
re.findall(".ai.", txt)

['rain', 'pain']

In [2206]:
# Return an empty list if no match was found
import re
txt = "The rain in Spain"
re.findall("Portugal", txt)

[]

### The search() Function
The `search()` function searches the string for a match.
- If there is more than one match, only the first occurrence of the match will be returned.

In [2207]:
# Search for the first white-space character in the string
# A raw string is used in the patternto include the backslash
import re
txt = "The rain in Spain"
re.search(r"\s", txt).start()

3

### The split() Function
The `split()` function returns a list where the string has been split at each match.

In [2208]:
# Split at each white-space character
# A raw string is used in the pattern to include the backslash
import re
txt = "The rain in Spain"
re.split(r"\s", txt)

['The', 'rain', 'in', 'Spain']

You can control the number of occurrences by specifying the `maxsplit` parameter.

In [2209]:
# Split the string only at the first occurrence
import re

txt = "The rain in Spain"
re.split(r"\s", txt, maxsplit=1)

['The', 'rain in Spain']

### The sub() Function
The `sub()` function replaces the matches with the text of your choice.

In [2210]:
# Replace every white-space character with the number 9
import re
txt = "The rain in Spain"
re.sub(r"\s", "-", txt)

'The-rain-in-Spain'

You can control the number of replacements by specifying the `count` parameter.

In [2211]:
# Replace the first 2 occurrences
import re
txt = "The rain in Spain"
re.sub(r"\s", "-", txt, count=2)

'The-rain-in Spain'

### Matching with groups
Text matching a pattern can be captured into a group by putting the match pattern between parentheses `()`.
The text matching each group can be accessed separately.

In [2212]:
# match each field of a date
import re
re.search(r"(\d{2})/(\d{2})/(\d{4})", "07/04/2022").groups()

('07', '04', '2022')

In [2213]:
# get month of a date
import re
re.search(r"(\d{2})/(\d{2})/(\d{4})", "07/04/2022").groups()[0]

'07'

In [2214]:
# get day of a date
import re
re.search(r"(\d{2})/(\d{2})/(\d{4})", "07/04/2022").groups()[1]

'04'

In [2215]:
# get year of a date
import re
re.search(r"(\d{2})/(\d{2})/(\d{4})", "07/04/2022").groups()[2]

'2022'

In [2216]:
# match each field of an email address
import re
re.search(r"(\w+)@(\w+)", "monty_python@gmail.com").groups()

('monty_python', 'gmail')

In [2217]:
# match each field of a web address
import re
re.search(r"https://www.(\w+)\.(\w+)", "https://www.google.com").groups()

('google', 'com')

In [2218]:
# match each field of an address
import re
re.search(r"(\w+), (\d+ .+), (.*), (\w{2}) (\d+)", "Googleplex, 1600 Amphitheatre Pkwy, Mountain View, CA 94043").groups()

('Googleplex', '1600 Amphitheatre Pkwy', 'Mountain View', 'CA', '94043')

# Placeholders
A dataset is the collection of the data being processed in a data analytics session.
The current values for all that data at any point in time is a *data state*.
The initial data state are all the available raw data.
This data is modified step by step as it is cleaned, analyzed, transformed, and visualized.
Placeholders, or *variables*, are needed to hold the data from step to step.
The placeholders are called *variables* because the data they hold may be changed from step to step.
This is distinct from mathematical variables, whose value never changes once it is associated with them.
The data state are the values contained by all existing variables.
When all steps are completed, the data is no longer needed and is released.
Any data needed for later must be saved before it is removed.

# Programs
A program automates the processing of data through data states, from the initial raw data until the last data state is handled and the session is over.
A program is a recipe and consists of a series of steps where some data is needed for each step, the data is modified, and the changed data results in a new data state. A recipe for making bread gives an example.

- initial data state: (flour, water, and yeast)
- make the dough
  - input: flour, water, and yeast
  - processing: mix and let rise
  - output: raw dough
  - data state: (raw dough)
- knead the dough
  - input: raw dough
  - processing: roll, press, and fold the dough
  - output: smooth dough
  - data state: (smooth dough)
- bake the dough:
  - input: smooth dough
  - processing: heat in the oven
  - output: bread
  - data state: (bread)

The initial data state is (flour, water, yeast) and the final data state is (bread).

In Python, the data state is stored in variables as the types of data and data structures above.
At each step, the data held by a variable may change,
or a variable may be assigned new data values, in which case it no longer needs the previous value and it is released (called *garbage collection*).
The data state is then the new set of values assigned to all variables.

## Program Statements
A program statement is responsible for each step of the processing that changes the data state.
Program statements may:
- assign data to a variable
- modify the data associated with a variable
- pass data to a function that uses the data to cause some external effect such as:
  - saving the data
  - creating a report from the data
  - visualizing the data
  - giving the data to a machine learning algorithm to predict future effects

Each statement modifies the data state.
The progression of a program as it is run is:
```
(initial data state)
program statement 1
(data state 1)
program statement 2
(data state 2)
...
last program statement
(all data is released and the program ends)
```

# Writing Python programs
Program statements, for our purposes, are written in Python and first some discussion of writing Python programs is needed.
Python programs are plain text files.
They are created in plain text editors such as:
- Notepad++ (Windows)
- TextEdit (MacOS)
- Sublime (Windows and MacOS)

Applications, such as PyCharm and Visual Studio Code, and websites such as Google Colab and w3schools, have built in text editors,
The files names usually end with “.py”.

## Python program structure
Comments are included to give text descriptions of what the code is doing in two ways:
- Text following "#" on one line is ignored through the end of the line
- Text between two single lines containing three double-quotes (“””) is ignored:
```
# this line is ignored

p = 42 # text after the "#" is ignored

“””
This is an ignored comment about the following program code.
“””
...
```
Long program lines can be continued on the next line, they are usually indented using tabs or spaces.
```
# A program line too long to fit on a page
foo = long_function_name(var_one, var_two,
                         var_three, var_four)
```
Occasionally a group of statements must be processed together under some statement, called a *block*.
These lines have exactly equal indentation (any size) to indicate processing them together:
```
# two lines run together if a condition is true
if (x < 3 and y == 4):
    z = x + y
    print(x, y, z)
```
Blank lines can be added anywhere between program lines to make it more readable.

At the beginning of files, programs usually indicate any standard libraries of code they use with “import”.
```
import os
import sys
# rest of the program
```

## Python program statements

Programs consist of a series of text lines containing statements that are processed from the beginning of the file to the end.
When all lines of the program have been processed, all values in the data state are released.
All data needs to have been processed by that point, and once it is released the computer is available to run another program.
These are various types of statements available.

### `print()`

One of the most basic statements is `print()`, which allows printing out the values associated with variables.
The format of the `print()` statement is:

`print(`*expression*`, `*expression*`, ...)`

`print()` prints out the values of the expressions on a single line, where each expression is separated by a space.


In [2219]:
print('the sum of ', 17, ' and ', 25, ' is ', 17+25)

the sum of  17  and  25  is  42


We have seen *escape sequences* where backslashes change the meaning of the following character in strings and regular expressions.
Two new ones insert a *tab*, or an indentation, or a *newline*, or a new line break, into a string.
| sequence | result |
| - | - |
| \n | Inserts a newline (line break) |
| \t | Inserts a horizontal tab (indentation) |


In [2220]:
print("\tstring indented with a tab")
print("string split\nonto two lines")

	string indented with a tab
string split
onto two lines


Besides that `print()` separates each expression with a space, it ordinarily ends each line on a new line by printing a newline `\n` at the end.
There is a way to change `print()` so it separates each expression and ends each line differently.

These can be changed by specifying two new options.

`print(`*expression*`, `*expression*`, ..., sep='`*string*`', end='`*string*`')`

- `sep=` lets you use another string besides a space to separate expressions.
- `end=` lets you end the line with something other than a new line.

In [2221]:
print("row", "row", "row your boat", sep=", ")
print("string on the same line as", end=": ")
print("this string")

row, row, row your boat
string on the same line as: this string


### Variable assignment statements
The most basic program statement assigns a value to a variable with `=`.
Variables are assigned to the value of *expressions*.
An expression is a combination of programming language variables, constants, operators, and functions that produce a single value.
Expressions follow the rules of standard mathematical statements.
Any place a variable or function appears in an expression, its value is substituted.

A variable initially is empty. It must be assigned a value before it is used, else it gives an error. This commonly occurs when a variable name is misspelled.

In [2222]:
# data state: ()
p = 42
# data state: (p, value 42)
print(p)

42


The data state at this point is:
- (p, value 42)

Any variable then can be used and have its value substituted in an expression in any following statement.

In [2223]:
# data state: ()
width = 2
# data state: (width = 2)
height = 21
# data state: (width = 2; height = 21)
area = width * height
# data state: (width = 2; height = 21; area = value 42)
print(area)

42


### Reassigning variables
After assigning a variable in a statement, the variable can be assigned a new value in a later statement.

In [2224]:
# data state: ()
width = 2
# data state: (width = 2)
height = 21
# data state: (width = 2; height = 21)
area = width * height
print(area)
# data state: (width = 2; height = 21; area = 42)
width = 4
# data state: (width = 4; height = 21; area = 42)
area = width * height
# data state: (width = 4; height = 21; area = 84)
print(area)

42
84


### Operator assignment shortcuts
A common variable assignment statement is to add to, substracts from, multiply by, or divide a variable with a value.

In [2225]:
p = 40
p = p + 2
print(p)
q = 50
q = q - 8
print(q)
r = 21
r = r * 2
print(r)
s = 126
s = s / 3
print(s)

42
42
42
42.0


Because this is so common, there are shortcut operators that are more concise.
| shortcut | operation |
| - | - |
| += | addition assignment |
| -= | substraction assignment |
| *= | multiplication assignment |
| /= | division assignment |

In [2226]:
p = 40
p += 2
print(p)
q = 50
q -= 8
print(q)
r = 21
r *= 2
print(r)
s = 126
s /= 3
print(s)

42
42
42
42.0


### Statements that modify lists
The following are operations that modify lists assigned to variables.

In [2227]:
# add a value to a list
a_list = ["string", 1.1, 42]
a_list.append(12)
print(a_list)

['string', 1.1, 42, 12]


In [2228]:
# add a list to the end of a list
a_list = ["string", 1.1, 42]
a_list.extend([12, "end"])
print(a_list)

['string', 1.1, 42, 12, 'end']


In [2229]:
# remove a value from a list
a_list = ["string", 1.1, 42]
a_list.remove(42)
print(a_list)

['string', 1.1]


In [2230]:
# add a value at a specific position
a_list = ["string", 1.1, 42]
a_list.insert(0, 3)
print(a_list)

[3, 'string', 1.1, 42]


In [2231]:
# delete a value at a specific position
a_list = ["string", 1.1, 42]
del a_list[1]
print(a_list)

['string', 42]


In [2232]:
# reverse a list
a_list = ["string", 1.1, 42]
a_list.reverse()
print(a_list)

[42, 1.1, 'string']


In [2233]:
# sort a list
a_list = [3, 12.1, 42, 10]
a_list.sort()
print(a_list)

[3, 10, 12.1, 42]


### Statements that modify sets
The following are operations that modify sets assigned to variables

In [2234]:
# add a value to set
a_set = {'apple', 'banana', 'orange', 'pear'}
a_set.add('grape')
print(a_set)

{'grape', 'pear', 'apple', 'orange', 'banana'}


In [2235]:
# remove a value from set
a_set = {'apple', 'banana', 'orange', 'pear'}
a_set.remove('orange')
print(a_set)

{'banana', 'pear', 'apple'}


In [2236]:
# remove all values from a set
a_set = {'apple', 'banana', 'orange', 'pear'}
a_set.clear()
print(a_set)

set()


### Statements that modify dictionaries
The following are operations that modify dictionaries assigned to variables

In [2237]:
# change a values in a dictionary
a_dict = {'leaf': 'green', 'sky': 'blue', 'apple': 'red'}
a_dict["apple"] = "yellow"
print(a_dict)

{'leaf': 'green', 'sky': 'blue', 'apple': 'yellow'}


In [2238]:
# add a value to a dictionary
a_dict = {'leaf': 'green', 'sky': 'blue', 'apple': 'red'}
a_dict["earth"] = "brown"
print(a_dict)

{'leaf': 'green', 'sky': 'blue', 'apple': 'red', 'earth': 'brown'}


In [2239]:
# remove a value from a dictionary
a_dict = {'leaf': 'green', 'sky': 'blue', 'apple': 'red'}
del a_dict["sky"]
print(a_dict)

{'leaf': 'green', 'apple': 'red'}


In [2240]:
# combine two dictionaries
a_dict = {'leaf': 'green', 'sky': 'blue', 'apple': 'red'}
another_dict = {"apple": "yellow", "sunset": "orange"}
a_dict.update(another_dict)
print(a_dict)

{'leaf': 'green', 'sky': 'blue', 'apple': 'yellow', 'sunset': 'orange'}


In [2241]:
# remove all pairs in a dictionary
a_dict = {'leaf': 'green', 'sky': 'blue', 'apple': 'red'}
a_dict.clear()
print(a_dict)

{}


## Statements that control statement processing order
The most powerful feature of programs is that they can process huge amounts of data and automatically make decisions based on data values.
Statements are normally processed from the top of the program to the end.
There are two statements that select whether a group of statements are to be processed:
- `if` statements that select whether to process a group of statements based on a condition.
- `match` statements that select a group of statements to process based on a value.

There are also two statements that allow processing groups of statements repeatedly:
- `while` statements repeat a group of statements while a condition is true.
- `for` statements that process a group of statements once for each data value in a list.

The statements being controlled by these are grouped by being indented with the exact same indentation for each statement as a *block*.
These statements in each block are then processed from first to last.

### `if` Statements
A block of statements are processed together if a condition is true with an `if` statement.
The condition follows the `if` and a colon `:` ends the line.

In [2242]:
x = 42
if x == 42:
    print('We have the Answer to the Ultimate Question of Life, the Universe, and Everything')

We have the Answer to the Ultimate Question of Life, the Universe, and Everything


One variation of an `if` statement allows one block of statements to be processed if the condition is true and another to be processed if the condition is false.
The statements to be processed if the condition are false follow an `else` statement that matches an `if` statement that has the same size indentation.
The `else` is followed by a colon `:`.

In [2243]:
x = 42
if x == 42:
    print('We have the Answer to the Ultimate Question of Life, the Universe, and Everything')
else:
    print('That is not quite the answer we were looking for')

We have the Answer to the Ultimate Question of Life, the Universe, and Everything


Another variation of an `if` statement allows multiple conditions to be tested one by one and process different groups of statements depending on the resuts.If there are multiple conditions to be tested, the first condition is with an `if` statement at the beginning, then rest of the conditions are with an `elif` statement.
The `elif` statement is followed by a colon `:`.
The statements following the `elif` have the same size indentation.
There can be a `else` statement after the last `elif` statement,
though that is not required

In [2244]:
x = 42
if x == 42:
    print('We have the Answer to the Ultimate Question of Life, the Universe, and Everything')
elif x == 41 or x == 43:
    print('Close, but not quite')
else:
    print('That is not quite the answer we were looking for')

We have the Answer to the Ultimate Question of Life, the Universe, and Everything


### `match` Statements
A match statement takes a value and processes the `case` blocks matching the value.
This works if the condition is that an expression is equal to a single value.

In [2245]:
word = "abc"
n = len(word)
match n:
    case 1:
        print(word, " has 1 character")
    case 2:
        print(word, " has 2 characters")
    case 3:
        print(word, " has 3 characters")

abc  has 3 characters


This is like an `elif` statement but is clearer and easier to read.

In [2246]:
word = "abc"
n = len(word)
if n == 1:
        print(word, " has 1 character")
elif n == 2:
        print(word, " has 2 characters")
elif n == 3:
        print(word, " has 3 characters")

abc  has 3 characters


There can be a the last `case` block with the value `_` that acts like an `else` if no other block matches.

In [2247]:
word = "abc"
n = len(word)
match n:
    case 1:
        print(word, " has 1 character")
    case 2:
        print(word, " has 2 characters")
    case 3:
        print(word, " has 3 characters")
    case _:
        print(word, " has", n, "characters")

abc  has 3 characters


### `while` Statements
Statements can be processed as long as a condition is true with the `while` statement.

In [2248]:
# print all even numbers up to 10
n = 2
while (n <= 10):
  print(n)
  n += 2

2
4
6
8
10


### `for` Statements
Statements can be processed for each value in a list with the `for` statement.

In [2249]:
# Measure some strings:
words = ['apple', 'pear', 'orange']
for word in words:
    print(word, len(word))

apple 5
pear 4
orange 6


A shortcut for a list of a set of numbers is the `range` function.
There are 3 versions of using `range`.
| code | meaning |
| - | - |
| `range(`*n*`)` | create a list from 0 to *n*-1 |
| `range(`*start*`, `*finish*`)` | create a list from *start* to *finish*-1 |
| `range(`*start*`, `*finish*`, `*step*`)` | create a list from *start* to *finish*-1 skipping by *step* |

In the version `range(`*start*`, `*finish*`, `*step*`)`,
the last number will be less than *finish-1* if the last *step* takes it past that value.

In [2250]:
# prints numbers from  0 to 4
for i in range(5):
    print(i)

0
1
2
3
4


In [2251]:
# prints numbers from 8 to 11
for i in range(8, 12):
    print(i)

8
9
10
11


In [2252]:
# prints odd numbers to 11
for i in range(1, 12, 2):
    print(i)

1
3
5
7
9
11


In [2253]:
# prints every 3th number under 12
for i in range(3, 12, 3):
    print(i)

3
6
9


A `for` loop can use `range` to process each list value using an *index*.

In [2254]:
# Measure some strings:
words = ['apple', 'pear', 'orange']
for i in range(len(words)):
    print(i, words[i], len(words[i]))

0 apple 5
1 pear 4
2 orange 6


### break and continue Statements
The break statement breaks out of the innermost enclosing for or while loop:

In [2255]:
# print squares of numbers until one is divisible by 3
for i in range(2, 10, 2):
  if i % 3 == 0:
     print(i, "is dividible by 3")
     break
  print(i)

2
4
6 is dividible by 3


The continue statement continues with the next iteration of the loop:

In [2256]:
# print squares of numbers not divisible by 3
for i in range(2, 10, 2):
  if i % 3 == 0:
      print(i, "is dividible by 3, skipping")
      continue
  else:
      print(i*i)


4
16
6 is dividible by 3, skipping
64


## Defining Functions
Functions collect blocks of statements that have one purpose.
Python has a variety of built-in functions such as `print()` that we used above.
You can define custom functions for any tasks you have that can make programs clearer and avoid repeating code.
Functions are defined with `def` followed by the function name and parentheses, ending with a colon `:`.
Then the statements of the function follow as an indented block as for conditional statements and loops.

In [2257]:
# print a greeting
def hello():
    print("hello")
hello()

hello


Functions can vary their processing using *parameters* passed when the functions is called.
The function call must specify *arguments* in the call to the function that matches all parameters defined.

In [2258]:
# print the cube of a number
def cube(n):
    print(n*n*n)
cube(8)

512


Functions can return values with the `return` statement.
When the return statement is processed, the value following it is returned as the function value.

In [2259]:
def cube(n):
    return n*n*n
print(cube(8))

512


#### Default Parameter Values
When a function is commonly called with a specific value for a parameter, a *default* value may be given to that parameter and the argument can be skipped in the call.

In [2260]:
# print a greeting
def hello(greeting="hello"):
    print(greeting)
hello("greetings")
hello()

greetings
hello


#### Keyword Parameters
Usually function arguments must be in the same order as the parameters where the function was defined.
The arguments can include the parameter name as a *keyword* in the call to the function so that the arguments can be in any order.

In [2261]:
# print a greeting with a name
def hello(greeting, name):
    print(greeting, name)
hello("greetings", "earthling")
hello(name="earthling", greeting="greetings")

greetings earthling
greetings earthling


### List Comprehensions
List *comprehensions* provide a concise way to create lists. It is common to produce lists when each element is the result of some operations applied to a sequence of values.

One way to create a list of squares with a `for` loop.

In [2262]:
# create a list of squares
squares = []
for x in range(10):
    squares.append(x**2)
print(squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


A list comprehension puts the `for` loop inside square brackets `[]` of a list where the form is:  
`[`*value* `for` *variable* `in` *list*`]`

In [2263]:
squares = [x**2 for x in range(10)]
print(squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


## A Python data science program to read in data and prepare it for processing
Here is a Python program to get data for US States (name, abbreviation, population, number of representatives, etc.) as a first step for data analysis.

In [2264]:
import requests

url = "https://gist.githubusercontent.com/kushnertodd/c7d54822e0fb946a71cecd8d0ba23d7a/raw/7372e07ce4fd295c070e4f3b362353814cc6df80/states.csv"
response = requests.get(url)

# Check if the request was successful
if response.status_code == 200:
    data = response.text
    print(data)

name,abbrev,admitted,population 2020,sq miles,sq km,reps
Alabama,AL,12/14/1819,5024279,52420,135767,7
Alaska,AK,1/3/1959,733391,665384,1723337,1
Arizona,AZ,2/14/1912,7151502,113990,295234,9
Arkansas,AR,6/15/1836,3011524,53179,137732,4
California,CA,9/9/1850,39538223,163695,423967,52
Colorado,CO,8/1/1876,5773714,104094,269601,8
Connecticut,CT,1/9/1788,3605944,5543,14357,5
Delaware,DE,12/7/1787,989948,2489,6446,1
Florida,FL,3/3/1845,21538187,65758,170312,28
Georgia,GA,1/2/1788,10711908,59425,153910,14
Hawaii,HI,8/21/1959,1455271,10932,28313,2
Idaho,ID,7/3/1890,1839106,83569,216443,2
Illinois,IL,12/3/1818,12812508,57914,149995,17
Indiana,IN,12/11/1816,6785528,36420,94326,9
Iowa,IA,12/28/1846,3190369,56273,145746,4
Kansas,KS,1/29/1861,2937880,82278,213100,4
Kentucky,KY,6/1/1792,4505836,40408,104656,6
Louisiana,LA,4/30/1812,4657757,52378,135659,6
Maine,ME,3/15/1820,1362359,35380,91633,2
Maryland,MD,4/28/1788,6177224,12406,32131,8
Massachusetts,MA,2/6/1788,7029917,10554,27336,9
Michigan,MI,1

This splits the data lines into lists making each field available individually.

In [2265]:
header = True
for line in data.splitlines():
  if not line.strip():
    print(f"blank line:    {line}")
  elif header:
    #print(f"header:        {line}")
    header = False
    header_fields = line.split('\t')
    print(f"header fields: {header_fields}")
  else:
    #print(f"processing:    {line}")
    state_fields = line.split('\t')
    print(f"state fields:  {state_fields}")

header fields: ['name,abbrev,admitted,population 2020,sq miles,sq km,reps']
state fields:  ['Alabama,AL,12/14/1819,5024279,52420,135767,7']
state fields:  ['Alaska,AK,1/3/1959,733391,665384,1723337,1']
state fields:  ['Arizona,AZ,2/14/1912,7151502,113990,295234,9']
state fields:  ['Arkansas,AR,6/15/1836,3011524,53179,137732,4']
state fields:  ['California,CA,9/9/1850,39538223,163695,423967,52']
state fields:  ['Colorado,CO,8/1/1876,5773714,104094,269601,8']
state fields:  ['Connecticut,CT,1/9/1788,3605944,5543,14357,5']
state fields:  ['Delaware,DE,12/7/1787,989948,2489,6446,1']
state fields:  ['Florida,FL,3/3/1845,21538187,65758,170312,28']
state fields:  ['Georgia,GA,1/2/1788,10711908,59425,153910,14']
state fields:  ['Hawaii,HI,8/21/1959,1455271,10932,28313,2']
state fields:  ['Idaho,ID,7/3/1890,1839106,83569,216443,2']
state fields:  ['Illinois,IL,12/3/1818,12812508,57914,149995,17']
state fields:  ['Indiana,IN,12/11/1816,6785528,36420,94326,9']
state fields:  ['Iowa,IA,12/28/1846,

In [2266]:
import requests

uncleaned_url = "https://raw.githubusercontent.com/kushnertodd/basic-data-science/refs/heads/main/data/Audible%20Dataset/audible_uncleaned.tsv"
uncleaned_response = requests.get(uncleaned_url)

In [2267]:
# Check if the request was successful
# if uncleaned_response.status_code == 200:
#     uncleaned_data = uncleaned_response.text
#     header = True
#     first = True
#     n = 0
#     for line in uncleaned_data.splitlines('\t'):
#       print(line)
#       if not line.strip():
#         print(f"blank line:    {line}")
#       elif header:
#         header = False
#         uncleaned_header_fields = line.split('\t')
#         print(f"header fields: {uncleaned_header_fields}")
#       else:
#         uncleaned_audible_fields = line.split('\t')
#         if (first):
#           first_uncleaned = uncleaned_audible_fields
#           first = False
#         print(f"uncleaned audible fields: {len(uncleaned_header_fields)} {uncleaned_audible_fields}")
#       n += 1
#       if (n > 10):
#         break

In [2268]:
'''
given the header fields are:
  ['name', 'author', 'narrator', 'time', 'releasedate', 'language', 'stars', 'price']
the field values for each record can be added to lists created with each of the fieldnames.
example:
  Geronimo Stilton #11 & #12,Writtenby:GeronimoStilton,Narratedby:BillLobely,2 hrs and 20 mins,04-08-08,English,5 out of 5 stars34 ratings,468.00
after splitting on ',' becomes
  ['Geronimo Stilton #11 & #12', 'Writtenby:GeronimoStilton', 'Narratedby:BillLobely', '2 hrs and 20 mins', '04-08-08', 'English', '5 out of 5 stars34 ratings', '468.00']
'''
header = True
uncleaned_names = []
uncleaned_authors = []
uncleaned_narrators = []
uncleaned_times = []
uncleaned_releasedates = []
uncleaned_languages = []
uncleaned_stars = []
uncleaned_prices = []
# Check if the request was successful
if uncleaned_response.status_code == 200:
    uncleaned_data = uncleaned_response.text
    for line in uncleaned_data.splitlines():
      if line.strip(): # line is not blank
        if header:     # skip header line
          header = False
        else:
          uncleaned_audible_fields = line.split('\t')
          uncleaned_names.append(uncleaned_audible_fields[0])
          uncleaned_authors.append(uncleaned_audible_fields[1])
          uncleaned_narrators.append(uncleaned_audible_fields[2])
          uncleaned_times.append(uncleaned_audible_fields[3])
          uncleaned_releasedates.append(uncleaned_audible_fields[4])
          uncleaned_languages.append(uncleaned_audible_fields[5])
          uncleaned_stars.append(uncleaned_audible_fields[6])
          uncleaned_prices.append(uncleaned_audible_fields[7])
    print('names        ', uncleaned_names [0:3])
    print('authors      ', uncleaned_authors [0:3])
    print('narrators    ', uncleaned_narrators [0:3])
    print('times        ', uncleaned_times [0:3])
    print('releasedates ', uncleaned_releasedates [0:3])
    print('languages    ', uncleaned_languages [0:3])
    print('stars        ', uncleaned_stars [0:3])
    print('prices       ', uncleaned_prices [0:3])

names         ['Geronimo Stilton #11 & #12', 'The Burning Maze', 'The Deep End']
authors       ['Writtenby:GeronimoStilton', 'Writtenby:RickRiordan', 'Writtenby:JeffKinney']
narrators     ['Narratedby:BillLobely', 'Narratedby:RobbieDaymond', 'Narratedby:DanRussell']
times         ['2 hrs and 20 mins', '13 hrs and 8 mins', '2 hrs and 3 mins']
releasedates  ['4/8/2008', '1/5/2018', '6/11/2020']
languages     ['English', 'English', 'English']
stars         ['5 out of 5 stars34 ratings', '4.5 out of 5 stars41 ratings', '4.5 out of 5 stars38 ratings']
prices        ['468', '820', '410']


In [2269]:
# prepare to clean the data
cleaned_names = []
cleaned_authors = []
cleaned_narrators = []
cleaned_times = []
cleaned_releasedates = []
cleaned_languages = []
cleaned_stars = []
cleaned_prices = []

#### `name` field
Examples of the `name` field:
```
Geronimo Stilton #11 & #12
The Burning Maze
The Deep End
```
The `name` field need not be changed.

In [2270]:
cleaned_names = []
for uncleaned_name in uncleaned_names:
  cleaned_names.append(uncleaned_name)
print('cleaned_names ', cleaned_names[0:3])

cleaned_names  ['Geronimo Stilton #11 & #12', 'The Burning Maze', 'The Deep End']


#### `author` field
Examples of the `author` field:
```
Writtenby:GeronimoStilton
Writtenby:RickRiordan
Writtenby:JeffKinney
```
The `author` field can have the unnecessary 'Writtenby:' prefix removed:
```
GeronimoStilton
RickRiordan
JeffKinney
```

In [2271]:
cleaned_authors = []
for uncleaned_author in uncleaned_authors:
  cleaned_author = uncleaned_author.replace('Writtenby:', '')
  cleaned_authors.append(cleaned_author)
print('cleaned_authors ', cleaned_authors[0:3])

cleaned_authors  ['GeronimoStilton', 'RickRiordan', 'JeffKinney']


#### `narrator` field  
Examples of the `narrator` field:
```
Narratedby:BillLobely
Narratedby:RobbieDaymond
Narratedby:DanRussell
```
The `narrator` field can have the unnecessary 'Narratedby:'prefix removed:
```
BillLobely
RobbieDaymond
DanRussell
```

In [2272]:
cleaned_narrators = []
for uncleaned_narrator in uncleaned_narrators:
  cleaned_narrator = uncleaned_narrator.replace('Narratedby:', '')
  cleaned_narrators.append(cleaned_narrator)
print('cleaned_narrators ', cleaned_narrators[0:3])

cleaned_narrators  ['BillLobely', 'RobbieDaymond', 'DanRussell']


#### `time` field
Examples of the `time` field:
```
2 hrs and 20 mins
13 hrs and 8 mins
2 hrs and 3 mins
```
The `time` field can converted to minutes:
```
140
788
123
```
First it is worth determining all the different variant formats of the time field.
We can convert the digits of each field to 'd' to make the format generic:
```
d hrs and dd mins
dd hrs and d mins
d hrs and d mins
```
We can then add the fields to a set which will contain a unique copy of each converted field.

In [2273]:
formatted_uncleaned_time_set = set()
first = 0
for uncleaned_time in uncleaned_times:
  # convert digits to 'd'
  formatted_uncleaned_time = re.sub(r'\d', 'd', uncleaned_time)
  formatted_uncleaned_time_set.add(formatted_uncleaned_time)
for formatted_uncleaned_time in formatted_uncleaned_time_set:
  print(formatted_uncleaned_time)

ddd hrs and dd mins
d hr
d mins
dd mins
dd hrs and d min
d hrs and d mins
Less than d minute
dd hrs and d mins
d hr and d min
dd hrs and dd mins
d hr and d mins
d hrs
d hrs and d min
dd hrs
d min
d hrs and dd mins
d hr and dd mins


In [2274]:
import re

def parse_duration_to_minutes(line):
    line = line.strip()

    # 1. Handle special case: "Less than d minute" -> 1 minute
    if re.search(r'Less than \d minute', line):
        return 1

    # 2. Extract hours using 'hrs?' (matches 'hr' or 'hrs')
    hours_match = re.search(r'(\d+)\s*hrs?', line)
    hours = int(hours_match.group(1)) if hours_match else 0

    # 3. Extract minutes using 'mins?' (matches 'min' or 'mins')
    mins_match = re.search(r'(\d+)\s*mins?', line)
    minutes = int(mins_match.group(1)) if mins_match else 0

    # 4. Calculate total minutes
    return (hours * 60) + minutes

# lines = [
#     "123 hrs and 45 mins",
#     "1 hr",
#     "22 mins",
#     "Less than 5 minute",
#     "2 hrs and 1 min",
#     "12 hrs"
# ]
#
# for text in lines:
#     total_mins = parse_duration_to_minutes(text)
#     print(f"'{text}' -> {total_mins} minutes")
cleaned_times = []
for uncleaned_time in uncleaned_times:
  cleaned_time = parse_duration_to_minutes(uncleaned_time)
  cleaned_times.append(cleaned_time)
print('cleaned_times ', cleaned_times[0:3])

cleaned_times  [140, 788, 123]


#### `releasedate` field
Examples of the `releasedate` field:
```
6/11/2020
5/10/2021
13-01-10
30-10-18
```
For the release data fields of the format `6/11/2020` and `30-10-18` it is best to determine which of the fields are month, day, and year.
Even though it is likely they are 'month/day/year' and 'day-month-year', it is best to be sure.
This is best done by determining the minimum and maximum for the three fields for each of these formats.
We can then convert them to one consistent format, say 'month/day/year`.

In [2275]:
import re

# parse all the dates like 6/11/2020
slashdate_field1_min = 99999
slashdate_field1_max = 0
slashdate_field2_min = 99999
slashdate_field2_max = 0
slashdate_field3_min = 99999
slashdate_field3_max = 0
for uncleaned_releasedate in uncleaned_releasedates:
   slash_match = re.search(r'(\d+)/(\d+)/(\d+)', uncleaned_releasedate)
   if slash_match:
    slashdate_field1 = int(slash_match.group(1))
    slashdate_field2 = int(slash_match.group(2))
    slashdate_field3 = int(slash_match.group(3))
    slashdate_field1_min = min(slashdate_field1_min, slashdate_field1)
    slashdate_field1_max = max(slashdate_field1_max, slashdate_field1)
    slashdate_field2_min = min(slashdate_field2_min, slashdate_field2)
    slashdate_field2_max = max(slashdate_field2_max, slashdate_field2)
    slashdate_field3_min = min(slashdate_field3_min, slashdate_field3)
    slashdate_field3_max = max(slashdate_field3_max, slashdate_field3)
print('slashdate_field1_min: ', slashdate_field1_min)
print('slashdate_field1_max: ', slashdate_field1_max)
print('slashdate_field2_min: ', slashdate_field2_min)
print('slashdate_field2_max: ', slashdate_field2_max)
print('slashdate_field3_min: ', slashdate_field3_min)
print('slashdate_field3_max: ', slashdate_field3_max)

slashdate_field1_min:  1
slashdate_field1_max:  12
slashdate_field2_min:  1
slashdate_field2_max:  12
slashdate_field3_min:  1999
slashdate_field3_max:  2024


Given field 1 and 2 ranges are both 1-12, we'll chalk this up to this being synthetic data and use our intuition that the dates are the usual US format "month/day/year".

Now repeat that for dates of the format '30-10-18'.

In [2276]:
import re

# parse all the dates like 30-10-18
hyphendate_field1_min = 99999
hyphendate_field1_max = 0
hyphendate_field2_min = 99999
hyphendate_field2_max = 0
hyphendate_field3_min = 99999
hyphendate_field3_max = 0
for uncleaned_releasedate in uncleaned_releasedates:
   hyphen_match = re.search(r'(\d+)-(\d+)-(\d+)', uncleaned_releasedate)
   if hyphen_match:
    hyphendate_field1 = int(hyphen_match.group(1))
    hyphendate_field2 = int(hyphen_match.group(2))
    hyphendate_field3 = int(hyphen_match.group(3))
    hyphendate_field1_min = min(hyphendate_field1_min, hyphendate_field1)
    hyphendate_field1_max = max(hyphendate_field1_max, hyphendate_field1)
    hyphendate_field2_min = min(hyphendate_field2_min, hyphendate_field2)
    hyphendate_field2_max = max(hyphendate_field2_max, hyphendate_field2)
    hyphendate_field3_min = min(hyphendate_field3_min, hyphendate_field3)
    hyphendate_field3_max = max(hyphendate_field3_max, hyphendate_field3)
print('hyphendate_field1_min: ', hyphendate_field1_min)
print('hyphendate_field1_max: ', hyphendate_field1_max)
print('hyphendate_field2_min: ', hyphendate_field2_min)
print('hyphendate_field2_max: ', hyphendate_field2_max)
print('hyphendate_field3_min: ', hyphendate_field3_min)
print('hyphendate_field3_max: ', hyphendate_field3_max)

hyphendate_field1_min:  13
hyphendate_field1_max:  31
hyphendate_field2_min:  1
hyphendate_field2_max:  12
hyphendate_field3_min:  0
hyphendate_field3_max:  99


It looks like those dates have the format 'day-month-year (2 digit)', where we assume the year is after 2000 (year = 2000 + "year (2 digit)").

Let's put this together.

In [2277]:
import re

cleaned_releasedates = []
for uncleaned_releasedate in uncleaned_releasedates:
  # detect dates like 6/11/2020, leave unchanged
  slash_match = re.search(r'(\d+)/(\d+)/(\d+)', uncleaned_releasedate)
  if slash_match:
    cleaned_releasedate = uncleaned_releasedate
    cleaned_releasedates.append(cleaned_releasedate)
  else:
    # detect dates like 30-10-18, convert to 10/30/2018
    hyphen_match = re.search(r'(\d+)-(\d+)-(\d+)', uncleaned_releasedate)
    if hyphen_match:
      hyphendate_field1 = int(hyphen_match.group(1))
      hyphendate_field2 = int(hyphen_match.group(2))
      hyphendate_field3 = int(hyphen_match.group(3))
      cleaned_releasedate = f'{hyphendate_field2}/{hyphendate_field1}/{2000+int(hyphendate_field3)}'
      cleaned_releasedates.append(cleaned_releasedate)
print('uncleaned_releasedates ', uncleaned_releasedates[0:10])
print('cleaned_releasedates ', cleaned_releasedates[0:10])

uncleaned_releasedates  ['4/8/2008', '1/5/2018', '6/11/2020', '5/10/2021', '13-01-10', '30-10-18', '25-11-14', '2/5/2017', '2/5/2017', '24-09-19']
cleaned_releasedates  ['4/8/2008', '1/5/2018', '6/11/2020', '5/10/2021', '1/13/2010', '10/30/2018', '11/25/2014', '2/5/2017', '2/5/2017', '9/24/2019']


#### `language` field
Examples of the `language` field:
```
English
English
English
```
The `language` field need not be changed.

for uncleaned_name in uncleaned_names:
  cleaned_names.append(uncleaned_name)

In [2278]:
cleaned_languages = []
for uncleaned_language in uncleaned_languages:
  cleaned_languages.append(uncleaned_language)
print('cleaned_languages ', cleaned_languages[0:3])

cleaned_languages  ['English', 'English', 'English']


#### `stars` field
Examples of the `stars` field:
```
5 out of 5 stars34 ratings
4.5 out of 5 stars41 ratings
4.5 out of 5 stars38 ratings
```
The `stars` field can converted to just include the rating.
```
5
4.5
4.5
```

In [2279]:
import re
cleaned_stars = []
for uncleaned_star in uncleaned_stars:
  star_match = re.search(r'(\d+) out of', uncleaned_star)
  if star_match:
    cleaned_star = star_match.group(1)
    cleaned_stars.append(cleaned_star)
  elif uncleaned_star == "Not rated yet":
    cleaned_stars.append("0")
  else:
    print(f'badly formatted stars: {uncleaned_star}')
print('cleaned_stars ', cleaned_stars[0:3])

cleaned_stars  ['5', '5', '5']


#### `price` field  
Examples of the `price` field:
```
233
820
1,256.00
```
The `price` field is a little confusing, there doesn't seem to be a clear way to interpret it.
For simplicity it can be assumed not to be necessary and can be dropped.

In [2280]:
import requests

url = "https://raw.githubusercontent.com/kushnertodd/basic-data-science/refs/heads/main/data/Audible%20Dataset/audible_cleaned.csv"
response = requests.get(url)

In [2281]:
# Check if the request was successful
if response.status_code == 200:
    cleaned_data = response.text
    #print(data)
    header = True
    first = True
    n = 0
    for line in cleaned_data.splitlines():
      print(line)
      if not line.strip():
        print(f"blank line:    {line}")
      elif header:
        #print(f"header:        {line}")
        header = False
        cleaned_header_fields = line.split('\t')
        print(f"header fields: {cleaned_header_fields}")
      else:
        #print(f"processing:    {line}")
        cleaned_audible_fields = line.split('\t')
        if (first):
          first_cleaned = cleaned_audible_fields
          first = False
        print(f"cleaned audible fields: {len(uncleaned_header_fields)} {cleaned_audible_fields}")
      n += 1
      if (n > 10):
        break

name,author,narrator,time,releasedate,language,stars,price,ratings
header fields: ['name,author,narrator,time,releasedate,language,stars,price,ratings']
Geronimo Stilton #11 & #12,GeronimoStilton,BillLobely,140,2008-04-08,English,5.0,468.0,34.0
cleaned audible fields: 8 ['Geronimo Stilton #11 & #12,GeronimoStilton,BillLobely,140,2008-04-08,English,5.0,468.0,34.0']
The Burning Maze,RickRiordan,RobbieDaymond,788,2018-01-05,English,4.5,820.0,41.0
cleaned audible fields: 8 ['The Burning Maze,RickRiordan,RobbieDaymond,788,2018-01-05,English,4.5,820.0,41.0']
The Deep End,JeffKinney,DanRussell,123,2020-06-11,English,4.5,410.0,38.0
cleaned audible fields: 8 ['The Deep End,JeffKinney,DanRussell,123,2020-06-11,English,4.5,410.0,38.0']
Daughter of the Deep,RickRiordan,SoneelaNankani,676,2021-05-10,English,4.5,615.0,12.0
cleaned audible fields: 8 ['Daughter of the Deep,RickRiordan,SoneelaNankani,676,2021-05-10,English,4.5,615.0,12.0']
"The Lightning Thief: Percy Jackson, Book 1",RickRiordan,JesseB

In [2282]:
n = len(cleaned_header_fields)
for i in range(1,n-1):
  if (i > 10):
    break
  print(f'{i:6} {cleaned_header_fields[i]:12} {first_uncleaned[i]:28} {first_cleaned[i]:30}')

Here are the various time formats ("[0-9]" becomes "d"):
```
Less than d minute
d hr
d hr and d min
d hr and d mins
d hr and dd mins
d hrs
d hrs and d min
d hrs and d mins
d hrs and dd mins
d min
d mins
dd hrs
dd hrs and d min
dd hrs and d mins
dd hrs and dd mins
dd mins
ddd hrs and dd mins
```
Here are the various releasedata formats ("[0-9]" becomes "d"):
```
d/d/dddd
d/dd/dddd
dd-dd-dd (format = "mon-day-year" -> "2000+year-mon-day")
dd/d/dddd
dd/dd/dddd
```

# Running Python on your own
There are a couple of options for running Python programs once you are ready.

## Running from the web
These are a few websites you can use to run Python interactively from the browser. Most of these are only sandboxes to test small bits of Python code. To develop useful Python programs you will probably need one of the options to run it locally below.

- [CodeChef](https://www.codechef.com/ide)
- [CompilerExplorer](https://godbolt.org)
- [OneCompiler](https://onecompiler.com/python)
- [Online Python](https://www.online-python.com/)
- [Try It Online](https://tio.run/#python3)
- [W3schools](https://www.w3schools.com/tryit/trycompiler.asp?filename=demo_python)

## Google Colab

Google Colab actually lets you create and store Python programs on the web, though because it runs remotely it is more limited for accessing data and less direct for producing production programs.  You can launch it from this website:

https://colab.research.google.com/

Google Colab projects allows you to create Python programs as [Jupyter](https://jupyter.org/) *notebooks*, with explanations combined with Python code. Jupyter is a standard for Data Science projects. This document was produced as a Google Colab project:

[Yet Another Data Science Python Tutorial](https://colab.research.google.com/github/kushnertodd/basic-data-science/blob/main/basic_data_science_python.ipynb)

There are instructions for running it on this webpage: [Basic Data Science: Google Colab](https://basicdatascience.substack.com/p/basic-data-science-google-colab). You can set up a Colab project and it will store the notebook on your Google Drive. For production Python projects where you have more control over the data you read and produce, you probably want to run it from your PC.

## Running from an PC application
The best option for running Python locally is using a PC application that is an Interactive Development Environment (IDE) for running Python.
These provide a user interface to create, maintain, and run Python programs
There is not much reason for using anything else for Windows or MacOS, though these only run on those environments,

### JetBrains PyCharm
PyCharm is the most complete option for running Python on the PC. You can set up the PyCharm environment as follows:
- Go to the PyCharm download page: https://www.jetbrains.com/pycharm/download
- Select Download and run the installer
- To run, launch PyCharm from the Start menu
You will find yourself at a main screen where you can create or open a Python program and run it interactively.
It lets you set up Python projects and manage multiple Python programs simultaneously.
It has help for instructions for how to use the interface.

### Visual Studio Code (Windows)
Visual Studio Code is Microsoft's version of a Python environment.
It actually supports a number of programming languages so is not specialized to Python.
You can set up the environment like this:
- Go to the VS Code download page: https://code.visualstudio.com/download
- Select Download and run the installer
- Alternatively, and probably easier, install Visual Studio Code from the Microsoft Store (latest version)
- To run, launch VS Code from the Start menu
When you create a project, you have to specify it is for Python.
The instructions can help you create and store programs.

## Running from the command line
You may find running Python programs from the command line is more generally available,
though it would require skill in using computer terminal apps.
Python programs can be run from the command line like this:
```
python program.py
```
Programs can then access data on the local filesystem which you can manage with PC tools.
Running Python programs in production can be easier using command line scripts rather than having to run an IDE, load the program, run from the interace, etc.

Here are install instructions for Windows and MacOS.

### Install on Windows
- Go to the Python download page: phttps://www.python.org/downloads/
- Select the latest Python 3 version
- Run the installer and select Add Python to PATH (e.g., Python 3.14)
- Alternatively, and probably easier, install from the Microsoft Store (latest version)
- Open a command prompt and run this to test the installation succeeded:
```
C:> python --version
Python 3.14
```
- An alternative is to launch the Python Shell by running IDLE from the Start menu

### Install on MacOS
- Open Applications > Utilities > Terminal
- Run:
```
% brew install python3
```
- Run this to test the installation:
```
% python --version
Python 3.14
```